### ISSUE TITLE
failure to build r@3.5.0%gcc@7.3.0 arch=cray-cnl6-mic_knl   @ theta.ALCF

### ISSUE THREAD

**Issue by llpesce** [spack/spack#8605]
Thanks a lot for your previous help. It was always timely and to the point.

All the dependencies built OK after tinkering a bit and moving building to the compute nodes (I will find out how to farm out testing and bootstrapping to the computes or working around them).

I could not get the configuration step of R to work because it reported:
...
checking whether ftn and cc agree on int and double... configure: WARNING: ftn and cc disagree on int and double
...
Which to me basically meant that the config ran just failed, not necessarily that the numbers didn't agree. Thanks to help from ALCF suffort, I got that step fixed by adding 
export CRAYPE_LINK_TYPE=dynamic
export PMI_NO_FORK=1
Configure then went through, but it died later during build with 
  >> 884    /opt/cray/pe/cce/8.7.0/binutils/x86_64/x86_64-pc-linux-gnu/bin/ld: cannot find -lAtpSigHandler
  >> 885    /opt/cray/pe/cce/8.7.0/binutils/x86_64/x86_64-pc-linux-gnu/bin/ld: cannot find -lAtpSigHCommData
  >> 886    collect2: error: ld returned 1 exit status
those are libraries that are part of the cray module atp
> find /opt/cray/pe/atp/2.1.1/ -name \*AtpSigH\*
/opt/cray/pe/atp/2.1.1/libApp/libAtpSigHandler.so
/opt/cray/pe/atp/2.1.1/libApp/libAtpSigHandler.so.0
/opt/cray/pe/atp/2.1.1/libApp/libAtpSigHCommData.a
/opt/cray/pe/atp/2.1.1/libApp/libAtpSigHandler.a
/opt/cray/pe/atp/2.1.1/libApp/libAtpSigHandler.so.0.0.0

I tried to add the module load atp and build --dirty, but it didn't help. 
I am trying to work around it, but since it seems that something is non functioning correctly, I am reporting it (I am of course looking forward to your reply and to use it)

Thanks a lot

**Comment by adamjstewart** [comment]
@wscullin

**Comment by llpesce** [comment]
Update. Adding:
 
export CRAYPE_LINK_TYPE=dynamic
export PMI_NO_FORK=1
export LIBRARY_PATH=/opt/cray/pe/atp/2.1.1/libApp/:$LIBRARY_PATH

and building --dirty
spack install  --dont-restage --dirty r %gcc@7.3.0

Seems to work. I say seems because the compilation takes too long and the development queue kicks me out.  I could not make it work cross compiling (because it uses its own python which is built for the knl nodes, presumably I can make it an double arch installation, but it seems an overkill).

**Comment by scheibelp** [comment]
Spack should normally set `CRAYPE_LINK_TYPE=dynamic` automatically, so I'm not sure why you had to do it manually.

> I tried to add the module load atp and build --dirty, but it didn't help.

Does the `atp` module set `LIBRARY_PATH`? If it only sets `LD_LIBRARY_PATH` then the compiler won't be able to use that to find libraries at build time.

Regarding `PMI_NO_FORK` - did they mention why that should be added?

(FYI, I won't have a chance to look at this more until 7/9)

**Comment by llpesce** [comment]
> Spack should normally set CRAYPE_LINK_TYPE=dynamic automatically, so I'm not sure why you had to do it manually.
> 
> I tried to add the module load atp and build --dirty, but it didn't help.
> 
> Does the atp module set LIBRARY_PATH? If it only sets LD_LIBRARY_PATH then the compiler won't be able to use that to find libraries at build time.
> 
> 

Makes sense, you are right. I assumed that the module would set up pkg-config or something like it, but I should have checked that it was the case.


> Regarding PMI_NO_FORK - did they mention why that should be added?
> 
> 

It’s a instruction to the demons handling the transfer of executables to the compute nodes. My understanding is that it prevents the demons from running one executable per node and then forking to another node on the XC 40. 
That makes it possible to run multiple executable without producing a failure.

That isn’t 100% checked and just the sense I could make out of it that satisfied my curiosity.

**Comment by llpesce** [comment]
Currently, the only trick I need to get to the end of the build is to

export LIBRARY_PATH=/opt/cray/pe/atp/2.1.1/libApp/:$LIBRARY_PATH

And 

spack install  --dirty r %%gcc@7.3.0


I need to do it otherwise 
     883    /gpfs/mira-home/lpesce/SPACK/spack/lib/spack/env/gcc/gcc -shared -fopenmp -L/usr/local/lib64 -o libRblas.so blas.o  cmplxblas.o -lAtpSigHandler -lAtpSigHCommData -lrca -lsci_gnu_71_mpi -lsci_gnu_71 -lgfortran -lquadmath -lpthread -lm
  >> 884    /opt/cray/pe/cce/8.7.0/binutils/x86_64/x86_64-pc-linux-gnu/bin/ld: cannot find -lAtpSigHandler
  >> 885    /opt/cray/pe/cce/8.7.0/binutils/x86_64/x86_64-pc-linux-gnu/bin/ld: cannot find -lAtpSigHCommData

It breaks anyway, but a little later and with

make[6]: Leaving directory '/gpfs/mira-home/lpesce/SPACK/build_stage/spack-stage/spack-stage-vrT83v/R-3.5.0/src/library/tools/src'
make[5]: Leaving directory '/gpfs/mira-home/lpesce/SPACK/build_stage/spack-stage/spack-stage-vrT83v/R-3.5.0/src/library/tools/src'
make[4]: Leaving directory '/gpfs/mira-home/lpesce/SPACK/build_stage/spack-stage/spack-stage-vrT83v/R-3.5.0/src/library/tools'
make[4]: Entering directory '/gpfs/mira-home/lpesce/SPACK/build_stage/spack-stage/spack-stage-vrT83v/R-3.5.0/src/library/tools'
installing 'sysdata.rda'
Fri Jul 20 20:50:40 2018: [PE_0]:_pmi_alps_sync:alps response not OKAY
Fri Jul 20 20:50:40 2018: [PE_0]:_pmi_init:_pmi_alps_sync failed -1

This latter I don’t think that it is a spack problem even if it might have been caused by LIBRARY_PATH. My hope here is that someone has seen this problem already.

**Comment by adamjstewart** [comment]
@llpesce is this still an issue? @becker33 knows a lot about Cray.
